### ISSUE TITLE
Possibility of `alloc::sync::Arc` in `thumbv6m-none-eabi` target

### ISSUE THREAD

**Issue by alisomay** [embassy-rs/embassy#1764]
`embassy-sync` provides many useful sync primitives to work with but I couldn't seem to find an atomic reference counting pointer there to act as a replacement for `Arc`. 

Most of the issues of integrating crates to my project originates from the absence of `alloc::sync` in `thumbv6m-none-eabi` and mostly the issue is usage of `Arc`. A tangible recent example is `wasmi`. 

With the usage of 
`portable-atomic = { version = "1.3", default-features = false, features = ["require-cas", "critical-section"] }`
the missing atomic instructions are polyfilled and the compiler does not complain anymore although `alloc::sync` problem is still there and it spreads to many transative dependencies. 

I'm ok with vendoring or use forked versions of dependencies for this project but I'm confused about where to start from.

My questions are:
- Do you plan to implement a drop in replacement for `Arc` in `embassy-sync`?
- Is it wise to start a journey to implement a version of `Arc` which could work in `thumbv6m-none-eabi`?
- Are there other options. 

If you think that it is wise to start implementing a replacement `Arc` to unlock the integration of some dependencies, I'd appreciate if you would give me a few key words or entry points of research to avoid pitfalls. 

Thank you 🙏

**Comment by alisomay** [comment]
By the way just to see if my idea works. I butchered the core Arc implementation of Rust and ported all atomics to `portable-atomic` and move all `alloc` based imports to `embedded-alloc` and removed the allocator api replacing everything with the `Global` allocator. I have a Frankenstein `Arc` which acts like the original and it compiles. 

Then I replaced `Arc`s in all problematic deps and transative deps following up replacing the Mutex's with the ones from `embassy-sync`. I could make `wasmi` compile for `thumbv6m-none-eabi`.

Then I tried to run a hello world module on a real RP2040. 

Works fine 😅 
At least now I know that it is possible. 
The right way to do it is another story though..
